### ISSUE TITLE
`#[allow(dead_code)]` ignores warnings for other items too

### ISSUE THREAD

**Issue by camelid** [rust-lang/rust#80856]
I believe this should give a warning for `hello`, but it does not:

```rust
fn hello() {}

#[allow(dead_code)]
fn goodbye() { hello() }
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ce0a955bca31b26ee17a299a5079eb71))

Errors:

```
   Compiling playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 1.64s
```

**Comment by FabianWolff** [comment]
I think this is the intended behavior:
https://github.com/rust-lang/rust/blob/3e1c75c6e25a4db968066bd2ef2dabc7c504d7ca/compiler/rustc_passes/src/dead.rs#L384-L391

Note that this is different from prefixing the item name with an underscore:
```rust
fn hello() {}
fn _goodbye() { hello() }
```
```
warning: function is never used: `hello`
 --> src/lib.rs:1:4
  |
1 | fn hello() {}
  |    ^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: 1 warning emitted
```
The implementations are [here](https://github.com/rust-lang/rust/blob/3e1c75c6e25a4db968066bd2ef2dabc7c504d7ca/compiler/rustc_passes/src/dead.rs#L407-L410) and [here](https://github.com/rust-lang/rust/blob/3e1c75c6e25a4db968066bd2ef2dabc7c504d7ca/compiler/rustc_passes/src/dead.rs#L580-L594), respectively. Note how items marked with `#[allow(dead_code)]` are traversed deeply (by adding them to the worklist), whereas the check for whether an item name starts with an underscore happens at the last minute before emitting the warning, i.e. does not affect warnings for other items.
